sdk/typespec/typespec_client_core/src/http/headers/mod.rs

```diff
@@ -9,42 +9,66 @@ mod microsoft;
 pub use common::*;
 pub use microsoft::*;
 
-use std::{borrow::Cow, fmt::Debug, str::FromStr};
+use std::{borrow::Cow, convert::Infallible, fmt::Debug, str::FromStr};
 use typespec::error::{Error, ErrorKind, ResultExt};
 
 /// A trait for converting a type into request headers.
 pub trait AsHeaders {
+    type Error;
     type Iter: Iterator<Item = (HeaderName, HeaderValue)>;
-    fn as_headers(&self) -> Self::Iter;
+
+    fn as_headers(&self) -> Result<Self::Iter, Self::Error>;
 }
 
 impl<T> AsHeaders for T
 where
     T: Header,
 {
+    type Error = Infallible;
     type Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>;
 
     /// Iterate over all the header name/value pairs.
-    fn as_headers(&self) -> Self::Iter {
-        vec![(self.name(), self.value())].into_iter()
+    fn as_headers(&self) -> Result<Self::Iter, Self::Error> {
+        Ok(vec![(self.name(), self.value())].into_iter())
     }
 }
 
 impl<T> AsHeaders for Option<T>
 where
     T: AsHeaders<Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>>,
 {
+    type Error = T::Error;
     type Iter = T::Iter;
 
     /// Iterate over all the header name/value pairs.
-    fn as_headers(&self) -> Self::Iter {
+    fn as_headers(&self) -> Result<Self::Iter, T::Error> {
         match self {
             Some(h) => h.as_headers(),
-            None => vec![].into_iter(),
+            None => Ok(vec![].into_iter()),
         }
     }
 }
 
+/// Extract a value from the [`Headers`] collection.
+///
+/// The [`FromHeaders::from_headers()`] method is usually used implicitly, through [`Headers::get()`] or [`Headers::get_optional()`].
+pub trait FromHeaders: Sized {
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Holds a list of the header names that [`FromHeaders::from_headers`] expects.
+    ///
+    /// Used by [`Headers::get()`] to generate an error if the headers are not present.
+    const HEADER_NAMES: &'static [&'static str] = &[];
```

Comment:
That's actually what I started with but then I was thinking that it would usually be a `const` so I made it a `const`. I hadn't considered object-safety though, so I appreciate that check! I'll go back to using a method.

Reply:
Actually, I don't think being object-safe is useful here. There are no methods on `FromHeaders` that receive `self` as a parameter, since it's entirely for constructing objects. If you had a `&dyn FromHeaders`, there's nothing you can call on it. In order to _make_ this object-safe, I'd have to apply `where Self: Sized` constraints to all the methods (and remove the `: Sized` constraint from the trait itself). Doing that excludes those methods from being called on trait objects and there's nothing left to call.

Having said that, I _do_ think a function is a more flexible option here.